add class DatasetArranger

CunliangGeng opened the pull request:
This is a big PR to implement the pipelines of data arranging, which enables the local and podp modes.

Arranging data means
- creating data folders in the `root_dir`
-  downloading dataset if needed (e.g. for podp mode)
-  validating dataset downloaded or provided by users

Basically, it means all steps needed to make data ready for loading. 

The pipelines of arranging data for different types of data are displayed in the diagram of #117.

To keep the data arranging workflow simple, we use fixed project directory structure (see #163) with fixed dir and file names (see `globals.py`).

To use nplinker, users are required to
- create a `root_dir` manually and use it as the root directory of the nplinker project
- provide a config file `nplinker.toml` and put it in the `root_dir` 

**Major changes**
- Added file `arranger.py` including the class `DatasetArranger ` and some validation functions, which implement the pipelines of arranging data

- Clean/remove/update some files to make the arrangers work (some may need further refactoring in future PRs)
    -  cleaned `runbigscape.py`
    - Deleted `downloader.py` and its tests, which is replaced by `DatasetArranger`
    - Updated `loader.py` and `nplinker.py` to use the `DatasetArranger`

- Added integration tests for the arranger (tests passed)
  - Created `nplinker_local_mode.toml`
  - Updated `tests/conftest.py` 
   - Updated `test_nplinker_local.py` to test the `local mode` 
 

Tests on podp mode also passed on my local machine. Due to the cost of running bigscape, the tests will be added to the codebase in next PRs.

CunliangGeng committed:
Create arranger.py

- Add class `DatasetArranger`
- Add dataset validation functions `validate_gnps`, `validate_antismash` and `validate_bigscape`

CunliangGeng committed:
Delete downloader.py and its tests

CunliangGeng committed:
Update loader.py

CunliangGeng committed:
Update runbigscape.py

-  remove function `podp_run_bigscape`
- updated function `run_bigscape`

CunliangGeng committed:
Update nplinker.py

CunliangGeng committed:
Update conftest.py

remove invalid steps

CunliangGeng committed:
Update test_nplinker_local.py

CunliangGeng committed:
Update conftest.py

CunliangGeng committed:
Create nplinker_local_mode.toml

CunliangGeng committed:
update config tests

CunliangGeng commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/215?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

Current dependencies on/for this PR:
* **#215** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/215?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#214** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/214?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#212** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/212?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#211** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/211?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#210** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/210?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#209** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/209?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#208** <a href="https://app.graphite.dev/github/pr/NPLinker/nplinker/208?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `dev`

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/215?utm_source=stack-comment).

gcroci2 commented on src/nplinker/arranger.py:
```diff
@@ -0,0 +1,446 @@
+            validate_gnps(globals.GNPS_DEFAULT_PATH)
+
+        if config.mode == "podp":
+            # set range 3 to ensure download can try 2 times and downloaded data is valid
```
This will actually loop 3 times, not 2, right? Why are multiple times needed?

CunliangGeng replied:
Good catch! It can loop at most 3 times. Multiple times are required to make sure the downloaded data is valid; if not, try to download it again (and again).  The downloaded data might be invalid in the case of e.g. losing connection to internet, nplinker process killed, etc.

I updated this part of code and related ones to fix a bug in the logics: after 3 times of loop, the data should be validated again. See the new commit please.

gcroci2 commented on src/nplinker/loader.py:
```diff
@@ -43,138 +36,24 @@
-            "DatasetLoader({}, {}, {})".format(self._root, self.dataset_id, self._remote_loading)
-        )
-
-    def __repr__(self):
```
Are you sure we don't want to implement this anymore?

CunliangGeng replied:
I do not get which part of code you're talking about. The loader.py still needs further refactoring. In this PR I just cleaned it to make the arranger work.

gcroci2 replied:
I meant the `__repr__ ` magic method

CunliangGeng replied:
It's still there, I did not remove it. Github does not render the changes correctly...

gcroci2 reviewed (approved):
Great refactor :D
- Just for me to know, did you use a tool for creating the md diagram displayed in #117?
- I haven't run local tests myself, should I?

CunliangGeng committed:
fix arranging logics for gnps, antismash and bigscape

CunliangGeng committed:
set tests run on one core

this is a workaround to solve the issues in `tests/conftest.py`: it copies example data for each process if multi-process test is enabled

CunliangGeng commented:
> Great refactor :D
> 
> * Just for me to know, did you use a tool for creating the md diagram displayed in [PODP mode and local data mode #117](https://github.com/NPLinker/nplinker/issues/117)?

I talked about that on the sprint meeting last week ;-)  Two ways:
1. github support it natively, if you check the [raw data of that issue](https://github.com/NPLinker/nplinker/issues/117) (by editting the issue), you will see how it works.
3. a very good live editor https://mermaid.live/ (see [the diagram](https://mermaid.live/edit#pako:eNqNVF1v2jAU_StXfg5VzBpWeNjUlom2Wrao0SZtgQc3MdRVYme2oWWU_z5_xAW2MjUPVnw_zj3H19cbVIqKohGa1-KxvCdSw-fbKZ9yMN-l4HO2-CSlkMV4zUlp9lA6I6xIzSqimeBAbcDMp4yJJl2C-TsW9U3Ra12YFfQ9BRNAQrp45LUgVRF-QEhYUE4l0TQCwiuQtBErCvSJKc34wmUDmxt7TVeU61lgf15MvmS5TdIsT8_zK5d-wRb5JWnpDHq9D3CxyYhScFzcx60Hu7Dhz18EPO-fivftR_yg6hkmm2vllDXmbCH7Os4CjF8nHVgEtShJ7cJMFt6MBVWdzcmqmPRCQ_4E73i8HPWByxG48rKqwxYEkCsn_VWQzuVAXJMgnOYkoEfQiqoNnPtv4Nzfr9f19dBlYI3v5r-sb_aUh8vhkrVcQ0OeoA-aNZb4zX7CTopT4l0pu2OLorEr9IDUj2Rt6naoESwVlQq40MZlxoJWoAW0UqyYaedOqr9CafzWO5TGOwlpjN86Ul2aPyPH_GVKbo_MAmnbmpXkrqbuzgdlYdJCT20jC3s9rbgHWmp4UKb4nNXdeGS-u_YqW-NhX7O9tmbxbmJDdAeBD6JdOzJcfPda6S46kMq1JIynRoJRpAq_NQ32e8-jZ6b915JJWvkieXp04vJ0RzNP41cL78dZgnmKi0n36Pwlx0AcElW5eXdKTauOqQLVGQJV0dqWkroDyP-lYFvGtD9dNUMRaqhsCKvMs7yxtabIRDZ0ikbmt6Jzsqz1FE351oSSpRb5mpdopOWSRmjZWugxIwtJmmCkFdNCpv6ldw9-hFrCfwrxEmK2aLRBT2iET-OT0wEe9ofv40Efx8OzCK2NeXh6kiT9Uzw4wwOzJMk2Qr8dAj7pY5wkMcaD5N0wjnGy_QO8nBFF))

I first used the live editor and then copied the raw code to github that renders it as diagram automatically.

> * I haven't run local tests myself, should I?

Not necessary for this PR.

CunliangGeng commented:
### Merge activity

* **Mar 5, 10:38 AM EST**: @CunliangGeng started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/NPLinker/nplinker/215).

gcroci2 commented:
> > Great refactor :D
> > 
> > * Just for me to know, did you use a tool for creating the md diagram displayed in [PODP mode and local data mode #117](https://github.com/NPLinker/nplinker/issues/117)?
> 
> I talked about that on the sprint meeting last week ;-) Two ways:
> 
> 1. github support it natively, if you check the [raw data of that issue](https://github.com/NPLinker/nplinker/issues/117) (by editting the issue), you will see how it works.
> 2. a very good live editor https://mermaid.live/ (see [the diagram](https://mermaid.live/edit#pako:eNqNVF1v2jAU_StXfg5VzBpWeNjUlom2Wrao0SZtgQc3MdRVYme2oWWU_z5_xAW2MjUPVnw_zj3H19cbVIqKohGa1-KxvCdSw-fbKZ9yMN-l4HO2-CSlkMV4zUlp9lA6I6xIzSqimeBAbcDMp4yJJl2C-TsW9U3Ra12YFfQ9BRNAQrp45LUgVRF-QEhYUE4l0TQCwiuQtBErCvSJKc34wmUDmxt7TVeU61lgf15MvmS5TdIsT8_zK5d-wRb5JWnpDHq9D3CxyYhScFzcx60Hu7Dhz18EPO-fivftR_yg6hkmm2vllDXmbCH7Os4CjF8nHVgEtShJ7cJMFt6MBVWdzcmqmPRCQ_4E73i8HPWByxG48rKqwxYEkCsn_VWQzuVAXJMgnOYkoEfQiqoNnPtv4Nzfr9f19dBlYI3v5r-sb_aUh8vhkrVcQ0OeoA-aNZb4zX7CTopT4l0pu2OLorEr9IDUj2Rt6naoESwVlQq40MZlxoJWoAW0UqyYaedOqr9CafzWO5TGOwlpjN86Ul2aPyPH_GVKbo_MAmnbmpXkrqbuzgdlYdJCT20jC3s9rbgHWmp4UKb4nNXdeGS-u_YqW-NhX7O9tmbxbmJDdAeBD6JdOzJcfPda6S46kMq1JIynRoJRpAq_NQ32e8-jZ6b915JJWvkieXp04vJ0RzNP41cL78dZgnmKi0n36Pwlx0AcElW5eXdKTauOqQLVGQJV0dqWkroDyP-lYFvGtD9dNUMRaqhsCKvMs7yxtabIRDZ0ikbmt6Jzsqz1FE351oSSpRb5mpdopOWSRmjZWugxIwtJmmCkFdNCpv6ldw9-hFrCfwrxEmK2aLRBT2iET-OT0wEe9ofv40Efx8OzCK2NeXh6kiT9Uzw4wwOzJMk2Qr8dAj7pY5wkMcaD5N0wjnGy_QO8nBFF))

Nice thanks I was looking for something like 2.
